Introduce a `createFrontendFeatureLoader`

davidfestal opened the pull request:
## Hey, I just made a Pull Request!

Introduced a `createFrontendFeatureLoader()` function, as well as a `FrontendFeatureLoader` interface, to gather several frontend plugins, modules or feature loaders in a single exported entrypoint and load them, possibly asynchronously.
This new feature, very similar to the `createBackendFeatureLoader()` already available on the backend, supersedes the previous `CreateAppFeatureLoader` type which has been deprecated.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Rugvip commented on packages/frontend-plugin-api/src/wiring/createFrontendFeatureLoader.ts:
```diff
@@ -0,0 +1,83 @@
+
+/** @public */
+export interface CreateFrontendFeatureLoaderOptions {
+  name: string;
```
If possible I'd prefer to skip the name and use an approach similar to this instead: https://github.com/backstage/backstage/blob/45e7b21f665f9095691886d402cf8d4d90c3b3bf/packages/backend-plugin-api/src/wiring/createBackendFeatureLoader.ts#L51

Just looking to avoid requiring users to name things if it's not really needed x)

davidfestal commented on packages/frontend-plugin-api/src/wiring/createFrontendFeatureLoader.ts:
```diff
@@ -0,0 +1,83 @@
+  name: string;
+  load(deps: { config: ConfigApi }): Promise<{
+    features: FrontendFeature[];
+  }>;
```
Done

davidfestal commented:
@Rugvip I implemented all the changes we discussed during the last Framework SIG call.
Do you think you could give a deeper look, and tell is something else is required to merge ?

Rugvip commented on packages/frontend-defaults/report.api.md:
```diff
@@ -32,6 +32,8 @@ export interface CreateAppOptions {
     config: ConfigApi;
   }>;
   // (undocumented)
+  featureLoaderRecursionDepth?: number;
```
Meh, I don't think this is worth an option, let's just hardcode it at somewhere around 5-10 and leave it at that?

davidfestal replied:
Quite useful for tests though :-)

davidfestal replied:
Which is why I made it optional

davidfestal replied:
Fixed.

Rugvip commented on packages/frontend-defaults/src/createApp.tsx:
```diff
@@ -87,29 +94,93 @@ export function createApp(options?: CreateAppOptions): {
+
+    // Separate deprecated CreateAppFeatureLoader elements from the frontend features,
+    // and manage the deprecated elements first.
+    const [providedFeatures, createAppFeatureLoaders] = (
+      options?.features ?? []
+    ).reduce<[FrontendFeature[], CreateAppFeatureLoader[]]>(
+      (acc, item) => {
+        if ('$$type' in item) {
+          acc[0].push(item);
+        } else {
+          acc[1].push(item);
+        }
+        return acc;
+      },
+      [[], []],
+    );
```
Bit of a nit, but I think a `let a; let b; for (...) { if (...) { a.push(...) } else { b.push(...) }` would be a bit easier to read here. Same thing below

davidfestal replied:
Fixed

Rugvip commented on packages/frontend-defaults/src/createApp.tsx:
```diff
@@ -87,29 +94,93 @@ export function createApp(options?: CreateAppOptions): {
+      loadedFeatures.push(...otherFeatures);
+      for (const featureLoader of featureLoaders) {
+        if (featureLoaderNames.some(l => l === featureLoader)) {
+          throw new Error(
+            `Added several instances of feature loader ${featureLoader.description}`,
+          );
```
I think a `continue` here would be fine tbh

davidfestal replied:
What would then be the option you would recommend to at least log a warning ?

Rugvip commented on packages/frontend-app-api/src/wiring/createSpecializedApp.tsx:
```diff
@@ -200,6 +202,15 @@ export function createSpecializedApp(options?: {
+  const featureLoaders = options?.features?.filter(
+    isInternalFrontendFeatureLoader,
+  );
+  if (featureLoaders && featureLoaders.length > 0) {
+    throw new Error(
+      `FrontendFeatureLoader is an async construct. It should be passed to createApp(), not to createSpecializedApp().`,
+    );
+  }
```
So this is the kind of issue I was alluding the regarding the new meaning of the `FrontendFeature` type. Tbh, I think we should keep `FrontendFeature` where it is and as it is, and make `createApp` accept `FrontendFeature | CreateAppFeatureLoader | FrontendFeatureLoader` instead. I at least can't think of any reason right now why that wouldn't be enough and I think it's better to avoid this confusion and keep it type safe.

davidfestal replied:
So I guess that the `loader` method of the `FrontendFeatureLoader` itself should then return `FrontendFeature | FrontendFeatureLoader` ?
Having a single Type for anything that can be loaded by the feature loader (including nested feature loaders themselves) is quite useful.
The same remark applies to the discovery `getAvailableFeatures()` method, which we expect to possibly return feature loaders as well.

davidfestal replied:
Is a change in the types returned by `FrontendFeatureLoader.loader` and `getAvailableFeatures ()` the direction you would recommend ?

davidfestal replied:
@Rugvip Please could you provide your thought here, before I update the PR and rebase it on top of https://github.com/backstage/backstage/pull/29058 ?

Rugvip replied:
yep, essentially any point in the code that requires a sync logic would use `sync`, while any point that can do async as well would have `sync | async`

Rugvip reviewed (commented):
Thanks! Couple more bits bits but then good to go. Leaving it to be a race with #29058 😅

davidfestal commented:
@Rugvip I rebased this PR on top of https://github.com/backstage/backstage/pull/29058, and removed the `FrontendFeatureLoader` from the `FrontendFeature` type, in order to restore type-safety related to sync / async application creation.
Since I had fixed all your other comments, that should hopefully be in a good shape for merge, which would allow me to come back to fixing the other review comment on the pending PR #28076 about *__dynamic frontend plugins__*.

davidfestal commented:
> @Rugvip I rebased this PR on top of #29058, and removed the `FrontendFeatureLoader` from the `FrontendFeature` type, in order to restore type-safety related to sync / async application creation. Since I had fixed all your other comments, that should hopefully be in a good shape for merge, which would allow me to come back to fixing the other review comment on the pending PR #28076 about _**dynamic frontend plugins**_.

cc @iamEAP

davidfestal commented on packages/frontend-defaults/src/discovery.ts:
```diff
@@ -57,6 +60,7 @@ function readPackageDetectionConfig(config: Config) {
  */
 export function discoverAvailableFeatures(config: Config): {
   features: FrontendFeature[];
+  featureLoaders?: FrontendFeatureLoader[];
```
Fixed

davidfestal commented on packages/frontend-defaults/src/resolution.ts:
```diff
@@ -16,30 +16,93 @@
+      if (isBackstageFeatureLoader(item)) {
+        featureLoaders.push(item);
+      } else {
+        loadedFeatures.push(item);
```
Fixed

davidfestal commented:
> Two minor bits and then we'll be good to merge 👍

@Rugvip The 2 minor bits are fixed.

davidfestal commented:
@Rugvip Any idea why the DCO check, which was successful when submitting this PR, is now back in "Expected" ?

Rugvip commented on packages/frontend-defaults/src/resolution.ts:
```diff
@@ -16,30 +16,98 @@
+
+    for (const featureOrLoader of featuresOrLoaders) {
+      if (isBackstageFeatureLoader(featureOrLoader)) {
+        if (alreadyMetFeatureLoaders.some(l => l === featureOrLoader)) {
```
`.includes` would work here instead, or even better make it a `Set` and use `.has`. Merging as is though as this is minor

Rugvip reviewed (approved):
Nice, thank you! 👍 

All good, let's :shipit:

Rugvip commented:
I see DCO on all commits so the check is just stuck, merging
